resources/providers/awslib/cluster_name_provider.go

```diff
@@ -0,0 +1,135 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package awslib
+
+import (
+	"context"
+	"fmt"
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/autoscaling"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	"k8s.io/klog/v2"
+	"regexp"
+)
+
+type EKSClusterNameProvider struct {
+}
+
+const (
+	asgPattern     = "^kubernetes.io/cluster/(.*)$"
+	clusterNameTag = "eks:cluster-name"
+)
+
+var (
+	asgCompiledRegex = regexp.MustCompile(asgPattern)
+)
+
+type ClusterNameProvider interface {
+	GetClusterName(ctx context.Context, cfg aws.Config, instanceId string) (string, error)
+}
+
+func (provider EKSClusterNameProvider) GetClusterName(ctx context.Context, cfg aws.Config, instanceId string) (string, error) {
+	// With EKS, there is no data source that can guarantee to return the cluster name.
+	// Therefore, we need to try multiple ways to find the cluster name.
+	// First, try to extract the cluster name from the instance tags.
+	// This is the most reliable way to find the cluster name.
+	// However, this method will work only on new EKS clusters.
+	// Therefore, if the tag was not found we will try to extract the cluster name from the autoscaling group.
+	clusterName, err := provider.getClusterNameFromInstanceTags(ctx, cfg, instanceId)
+	if err != nil {
+		return "", fmt.Errorf("failed to get cluster name from the instance tags: %v", err)
+	}
+	if clusterName != "" {
+		return clusterName, nil
+	}
+	clusterName, err = provider.getClusterNameFromAutoscalingGroup(ctx, cfg, instanceId)
+	if err != nil {
+		return "", fmt.Errorf("failed to get cluster name from the Auto-scaling group: %v", err)
+	}
+
+	return clusterName, nil
+}
+
+func (provider EKSClusterNameProvider) getClusterNameFromInstanceTags(ctx context.Context, cfg aws.Config, instanceId string) (string, error) {
+	input := &ec2.DescribeInstancesInput{
+		InstanceIds: []string{instanceId},
+	}
+	svc := ec2.NewFromConfig(cfg)
+
+	for {
+		r, err := svc.DescribeInstances(ctx, input)
+		if err != nil {
+			return "", fmt.Errorf("failed to describe instance required for cluster name detection: %v", err)
+		}
+
+		// Look for the cluster name tag in the instance tags
+		for _, reservation := range r.Reservations {
+			for _, instance := range reservation.Instances {
+				if *instance.InstanceId == instanceId {
+					for _, tag := range instance.Tags {
+						if *tag.Key == clusterNameTag {
+							return *tag.Value, nil
+						}
+					}
+				}
+			}
+		}
+
+		if r.NextToken == nil {
+			break
+		}
+		input.NextToken = r.NextToken
+	}
+	return "", nil
+}
+
+func (provider EKSClusterNameProvider) getClusterNameFromAutoscalingGroup(ctx context.Context, cfg aws.Config, instanceId string) (string, error) {
```

Comment:
Looks like this was already implemented here: https://github.com/elastic/k8smd/blob/eb91ee249b9503d2c0d56ffbc523f11479fe9a00/pkg/cloudmetadata/provider_aws.go#L217-L252
Is there any reason not to reuse the existing code?

I also saw that there are more methods of getting the cluster name there that we haven't implemented, would like to understand the reasoning behind that.

I'm still wondering if there is a more efficient way of getting the cluster name though.

Reply:
So I made a lot of investigation, and I indeed relied on this example.

The project you refer to is not currently used in our Cloudbeat, and we will need to add it.
And then we will have to make this method public in order to use it.
Does it really worth it? adding another dependency? I don't know - but I guess it worth considering it

>I also saw that there are more methods of getting the cluster name there that we haven't implemented, would like to understand the reasoning behind that.

The other methods are not relevant they look for a certain pattern that is relevant only for EKS that resides in Elastic - you can see the regex.

Anyway, I think as you suggest on slack we can't use this code as it is now.
We will have to limit the search in scale groups to a certain number of rounds, we won't want cloudbeat to look on all customer available clusters, so I guess that if we will change this code there is no need to import it right?

Reply:
There is always the option of contributing to `k8smd` if it suits their agenda as well.
It is a package that is being developed within Elastic and there shouldn't be much friction doing that.